### PR TITLE
Candid bakery

### PR DESCRIPTION
**Pull request by sergiusens**
- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----
Depends on #3472

### CODE REVIEW

File: snapcraft/storeapi/_dashboard_api.py [hunk 1]
```diff
@@ -49,18 +49,32 @@ def _request(self, method: str, urlpath: str, **kwargs) -> requests.Response:
...
+            data["channels"] = channels
         if expires is not None:
-            data.update({"expires": expires})
-        headers = {"Accept": "application/json"}
-        response = self.post(
-            "/dev/api/acl/", json=data, headers=headers, auth_header=False
-        )
+            data["expires"] = expires
+
+        headers = {"Content--Type": "application/json", "Accept": "application/json"}
```
Comment (sergiusens) [inline snapcraft/storeapi/_dashboard_api.py:69]:
```suggestion
        headers = {"Content-Type": "application/json", "Accept": "application/json"}
```

**Comment by cjp256** [comment]
LGTM, but I'd probably drop the experimental flag if it's supposed to work and nothing is expected to change.

**Comment by sergiusens** [comment]
Chris Patterson writes:

> LGTM, but I'd probably drop the experimental flag if it's supposed to
> work and nothing is expected to change.

Experimental flag is because some code paths might not work on the store
side, so we need some usage to make sure everything works ok.

File: snapcraft/storeapi/http_clients/_candid_client.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+from snapcraft.storeapi import constants
+from . import agent, _config, _http_client
+
+
+class CandidConfig(_config.Config):
+    """Hold configuration options in sections.
+
+    There can be two sections for the sso related credentials: production and
+    staging. This is governed by the UBUNTU_ONE_SSO_URL environment
```
Comment (sergiusens) [inline snapcraft/storeapi/http_clients/_candid_client.py:21]:
```suggestion
    staging. This is governed by the STORE_DASHBOARD_URL environment
```

File: tests/unit/store/http_client/test_candid_client.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+
+def test_login_discharge_macaroon(candid_client, snapcraft_macaroon):
+    candid_client.request
+    candid_client.login(macaroon=snapcraft_macaroon)
+
+    assert candid_client.has_credentials() is True
+    assert candid_client._macaroon == snapcraft_macaroon
+    assert candid_client._auth == (
+        "W3siaWRlbnRpZmllciI6ICIiLCAic2lnbmF0dXJlIjogImQ5NTMzNDYxZDc4MzVlNDg1MWM"
```
Comment (cjp256) [inline tests/unit/store/http_client/test_candid_client.py:113]: Can this be done in a way to show how this magic string is derived? It's not obvious when looking at the tests.

File: tests/spread/general/store/task.yaml [hunk 4]
```diff
@@ -50,10 +51,18 @@ execute: |
...
 
+  # Logout is idempotent
+  snapcraft logout
+
   # Login
   set +x
-  echo "${SNAP_STORE_MACAROON}" | snapcraft login --with -
+  echo "${SNAP_STORE_MACAROON}" > login
   set -x
+  if [ "${SPREAD_VARIANT}" = "CANDID" ]; then
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply tests/spread/general/store/task.yaml:61]: I read the source :1st_place_medal:

File: tests/unit/store/http_client/test_candid_client.py (hunk 3 above)
Comment (sergiusens) [reply tests/unit/store/http_client/test_candid_client.py:113]: I can add a code comment, it's not magic

**Review by cjp256 — APPROVED** [review]

**Pull request merged**